tfhe/src/integer/server_key/radix_parallel/ilog2.rs

```diff
@@ -0,0 +1,734 @@
+                ct
+            })
+            .collect::<Vec<_>>();
+
+        self.unchecked_sum_ciphertexts_vec_parallelized(cts)
+            .expect("internal error, empty ciphertext count")
+    }
+
+    //==============================================================================================
+    //  Unchecked
+    //==============================================================================================
+
+    pub fn unchecked_trailing_zeros<T>(&self, ct: &T) -> RadixCiphertext
+    where
+        T: IntegerRadixCiphertext,
+    {
+        self.count_consecutive_bits(ct, Direction::Trailing, BitValue::Zero)
+    }
+
+    pub fn unchecked_trailing_ones<T>(&self, ct: &T) -> RadixCiphertext
+    where
+        T: IntegerRadixCiphertext,
+    {
+        self.count_consecutive_bits(ct, Direction::Trailing, BitValue::One)
+    }
+
+    pub fn unchecked_leading_zeros<T>(&self, ct: &T) -> RadixCiphertext
+    where
+        T: IntegerRadixCiphertext,
+    {
+        self.count_consecutive_bits(ct, Direction::Leading, BitValue::Zero)
+    }
+
+    pub fn unchecked_leading_ones<T>(&self, ct: &T) -> RadixCiphertext
+    where
+        T: IntegerRadixCiphertext,
+    {
+        self.count_consecutive_bits(ct, Direction::Leading, BitValue::One)
+    }
+
+    //==============================================================================================
+    //  Smart
+    //==============================================================================================
+
+    pub fn smart_trailing_zeros<T>(&self, ct: &mut T) -> RadixCiphertext
+    where
+        T: IntegerRadixCiphertext,
+    {
+        if !ct.block_carries_are_empty() {
+            self.full_propagate_parallelized(ct);
+        }
+
+        self.unchecked_trailing_zeros(ct)
+    }
+
+    pub fn smart_trailing_ones<T>(&self, ct: &mut T) -> RadixCiphertext
+    where
+        T: IntegerRadixCiphertext,
+    {
+        if !ct.block_carries_are_empty() {
+            self.full_propagate_parallelized(ct);
+        }
+
+        self.unchecked_trailing_ones(ct)
+    }
+
+    pub fn smart_leading_zeros<T>(&self, ct: &mut T) -> RadixCiphertext
+    where
+        T: IntegerRadixCiphertext,
+    {
+        if !ct.block_carries_are_empty() {
+            self.full_propagate_parallelized(ct);
+        }
+
+        self.unchecked_leading_zeros(ct)
+    }
+
+    pub fn smart_leading_ones<T>(&self, ct: &mut T) -> RadixCiphertext
+    where
+        T: IntegerRadixCiphertext,
+    {
+        if !ct.block_carries_are_empty() {
+            self.full_propagate_parallelized(ct);
+        }
+
+        self.unchecked_leading_ones(ct)
+    }
+
+    //==============================================================================================
+    //  Default
+    //==============================================================================================
+
+    /// Returns the number of trailing zeros in the binary representation of `ct`
+    ///
+    /// # Example
```

Comment:
add the standard blabla for the default ops ? I think we have it somewhere else

Reply:
Do people even read these ?

Reply:
no idea